beacon-chain/state/state-native/validator_index_cache.go

```diff
@@ -0,0 +1,97 @@
+// It includes a lastFinalizedIndex to track updates up to the last finalized validator index,
+// and uses a mutex for concurrent read/write access to the cache.
+type validatorIndexCache struct {
+	indexMap           map[[fieldparams.BLSPubkeyLength]byte]primitives.ValidatorIndex // Maps BLS public keys to validator indices.
+	lastFinalizedIndex int                                                             // Index of the last finalized validator to avoid reading validators before this point.
+	mutex              sync.RWMutex
+}
+
+// newValidatorIndexCache initializes a new validator index cache with an empty index map.
+func newValidatorIndexCache() *validatorIndexCache {
+	return &validatorIndexCache{
+		indexMap: make(map[[fieldparams.BLSPubkeyLength]byte]primitives.ValidatorIndex),
+	}
+}
+
+// getValidatorIndex retrieves the validator index for a given public key from the cache.
+// If the public key is not found in the cache, it searches through the state starting from the last finalized index.
+func (b *BeaconState) getValidatorIndex(pubKey [fieldparams.BLSPubkeyLength]byte) (primitives.ValidatorIndex, bool) {
+	b.validatorIndexCache.mutex.RLock()
+	index, found := b.validatorIndexCache.indexMap[pubKey]
+	b.validatorIndexCache.mutex.RUnlock()
+	if found {
+		return index, true
+	}
+
+	finalizedIndex := b.validatorIndexCache.lastFinalizedIndex
+	pubKeys, err := b.publicKeysSinceIndexReadOnly(finalizedIndex)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to get public keys since last finalized index %d", finalizedIndex)
+		return 0, false
+	}
+	for i, key := range pubKeys {
+		if key == pubKey {
+			index := primitives.ValidatorIndex(finalizedIndex + i)
+			return index, true
+		}
+	}
+	return 0, false
+}
+
+// saveValidatorIndices updates the validator index cache with new indices.
+// It processes validator indices starting after the last finalized index and updates the tracker.
+func (b *BeaconState) saveValidatorIndices() {
+	b.validatorIndexCache.mutex.Lock()
+	defer b.validatorIndexCache.mutex.Unlock()
+
+	finalizedIndex := b.validatorIndexCache.lastFinalizedIndex
+	pubKeys, err := b.publicKeysSinceIndexReadOnly(finalizedIndex)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to retrieve public keys starting from the last finalized index %d", finalizedIndex)
+		return
+	}
+	for i, key := range pubKeys {
+		b.validatorIndexCache.indexMap[key] = primitives.ValidatorIndex(finalizedIndex + i)
+	}
+	b.validatorIndexCache.lastFinalizedIndex += len(pubKeys)
+}
+
+// publicKeysSinceIndexReadOnly constructs a list of validator public keys starting from a specified index.
+// The indices in the returned list correspond to their respective validator indices in the state.
+// It returns an error if the specified index is out of bounds. This function is read-only and does not use locks.
+func (b *BeaconState) publicKeysSinceIndexReadOnly(index int) ([][fieldparams.BLSPubkeyLength]byte, error) {
```

Comment:
I was going to say it would be better for this to return a slice of validators, which would be super cheap if we had normal slices, but I suppose mv slice messes that up. Can mv slice return a go slice without making a bunch of copies?

Even it if has to copy validator pointers, that's still going to be a lot less than copying all the keys.

Another option is to design this as a method like `findValidator` that takes a callback ("query") and iterates through the validators until the callback returns true, then return the matching validator. Or similar idea but it iterates until the callback tells it to stop, and then the callback grabs the validator it wants via enclosure.

Reply:
Re: pointer, there is `ValidatorsReadOnly() []state.ReadOnlyValidator` which returns a list of 
```
type readOnlyValidator struct {
	validator *ethpb.Validator
}
```
Even then, 1M read-only validators (pointers) seem to be a lot for every deposit. I think it'd still make sense to implement `ValidatorsSinceIndex(index int)`. Do you agree with that?

Reply:
Yes I thought the point about only iterating validators since the index was implied - `ValidatorsSinceIndex(index int)` is actually what I had in mind :)

Reply:
Thanks. Updated the code to do that : )
Unfortunately, the read-only validator public key getter also requires a copy. I added a slice version
```go
func (v readOnlyValidator) PublicKey() [fieldparams.BLSPubkeyLength]byte {
	var pubkey [fieldparams.BLSPubkeyLength]byte
	copy(pubkey[:], v.validator.PublicKey)
	return pubkey
}
```